src/librustc_data_structures/smallvec.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! A vector type that allows storing small values directly on the stack.
+
+use std::iter::FromIterator;
+use std::ptr::drop_in_place;
+use std::ops::{Deref, DerefMut};
+use std::slice;
+use std::fmt;
+
+pub struct SmallVec8<T> {
+    count: usize,
+    values: [ManuallyDrop<T>; 8]
+}
+
+impl<T> SmallVec8<T> {
+    fn new() -> Self {
+        SmallVec8 {
+            count: 0,
+            values: [
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+            ],
+        }
+    }
+}
+
+impl<T> fmt::Debug for SmallVec8<T>
+    where T: fmt::Debug {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self[..].fmt(f)
+    }
+}
+
+impl<T> Deref for SmallVec8<T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        unsafe {
+            slice::from_raw_parts(self.values.as_ptr() as *const T, self.count)
+        }
+    }
+}
+
+impl<T> DerefMut for SmallVec8<T> {
+    fn deref_mut(&mut self) -> &mut [T] {
+        unsafe {
+            slice::from_raw_parts_mut(self.values.as_mut_ptr() as *mut T, self.count)
+        }
+    }
+}
+
+// FIXME: This should use repr(transparent) from rust-lang/rfcs#1758.
+#[allow(unions_with_drop_fields)]
+union ManuallyDrop<T> {
+    value: T,
+    #[allow(drop)]
+    empty: (),
+}
+
+impl<T> Drop for SmallVec8<T> {
+    fn drop(&mut self) {
+        unsafe {
+            drop_in_place(&mut self.values[..self.count]);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum MaybeSmallVec8<T> {
+    Small(SmallVec8<T>),
+    Large(Vec<T>)
+}
+
+impl<T> Deref for MaybeSmallVec8<T> {
```

Comment:
`MaybeSmallVec8` currently has very few methods defined on it. If it's going to get used at all widely that will need to change. I guess you just implemented the ones you needed for the other commits in this PR?

In contrast, `SmallVector` has quite a few methods implemented, but still not the full range that `Vec` has. E.g. I was working on a patch just yesterday where I used `SmallVector` and I had to add `as_mut_slice`.

Which leads to a more general point: is this type intended to replace `SmallVector`? Having two such types in the codebase seems silly. I know of numerous places where `Vec` is currently used but a small vector would be better, so having a good general-purpose small vector is a good thing. (I have three separate patches for which I haven't yet filed PRs where I currently am using `SmallVector` to get perf improvements.)

If we wanted the best of both worlds:
- It should be in librustc_data_structures. (`SmallVector` is in libsyntax, which isn't a great spot for it if it's going to get wider usage.)
- It should have an inline capacity of 2 or 4. 1 is small enough to miss some interesting cases; 8 is too big.
- It should support a wide range of methods, as `SmallVector` currently does.